AK/UFixedBigInt.h

```diff
@@ -17,474 +18,378 @@
 
 template<typename T>
-struct NumericLimits<UFixedBigInt<T>> {
-    static constexpr UFixedBigInt<T> min() { return 0; }
-    static constexpr UFixedBigInt<T> max() { return { NumericLimits<T>::max(), NumericLimits<T>::max() }; }
+concept NotBuiltInUFixedInt = UFixedInt<T> && !BuiltInUFixedInt<T>;
+
+// FIXME: This breaks formatting
+// template<typename T>
+// constexpr inline bool Detail::IsIntegral<UFixedBigInt<T>> = true;
+
+template<size_t bit_size>
+constexpr inline bool Detail::IsUnsigned<UFixedBigInt<bit_size>> = true;
+template<size_t bit_size>
+constexpr inline bool Detail::IsSigned<UFixedBigInt<bit_size>> = false;
+
+template<size_t bit_size>
+struct NumericLimits<UFixedBigInt<bit_size>> {
+    using T = UFixedBigInt<bit_size>;
+
+    static constexpr T min() { return T {}; }
+    static constexpr T max() { return --T {}; }
     static constexpr bool is_signed() { return false; }
 };
 
-template<Unsigned T>
-struct UFixedBigIntMultiplicationResult {
-    T low;
-    T high;
-};
+// ===== UFixedBigInt itself =====
+template<size_t bit_size>
+constexpr auto get_storage_of(UFixedBigInt<bit_size> const& value) { return value.m_data; }
+
+template<size_t bit_size, typename Storage>
+class UFixedBigInt {
+private:
+    template<size_t other_bit_size, typename OtherStorage>
+    friend class UFixedBigInt;
+
+    constexpr UFixedBigInt(ForOverwriteTag) { }
+
+    using RawLowUFixedBigInt = UFixedBigInt<bit_size / 2>;
+    using RawHighUFixedBigInt = UFixedBigInt<(bit_size + 1) / 2>;
+    using LowUFixedBigInt = Conditional<bit_size / 2 == 64, u64, RawLowUFixedBigInt>;
+    using HighUFixedBigInt = Conditional<(bit_size + 1) / 2 == 64, u64, RawHighUFixedBigInt>;
 
-template<typename T>
-requires(sizeof(T) >= sizeof(u64) && IsUnsigned<T>) class UFixedBigInt {
 public:
-    using R = UFixedBigInt<T>;
+    // TODO: make private
+    Storage m_data;
 
-    constexpr UFixedBigInt() = default;
-    template<Unsigned U>
-    requires(sizeof(T) >= sizeof(U)) constexpr UFixedBigInt(U low)
-        : m_low(low)
-        , m_high(0u)
+    constexpr UFixedBigInt()
+        : UFixedBigInt(0u)
     {
     }
-    template<Unsigned U, Unsigned U2>
-    requires(sizeof(T) >= sizeof(U) && sizeof(T) >= sizeof(U2)) constexpr UFixedBigInt(U low, U2 high)
-        : m_low(low)
-        , m_high(high)
+
+    template<UFixedInt T>
+    requires(sizeof(T) <= sizeof(Storage)) constexpr UFixedBigInt(const T& value)
     {
+        storage_copy(get_storage_of(value), m_data);
     }
-    constexpr T& low()
+
+    template<UFixedInt T>
+    requires(sizeof(T) > sizeof(Storage)) explicit constexpr UFixedBigInt(const T& value)
     {
-        return m_low;
+        storage_copy(get_storage_of(value), m_data);
     }
-    constexpr const T& low() const
+
+    constexpr UFixedBigInt(LowUFixedBigInt low, HighUFixedBigInt high)
     {
-        return m_low;
+        *this = high;
+        *this <<= bit_size / 2;
+        *this |= low;
```

Comment:
Would be nice if this could just be done with bit-bliting, ie no shifts

Reply:
Well, now I don't have any high and low parts, so the function is kinda artificial, existing only for backward-compatibility. Nonetheless, I might split the integer by some word boundary, then it will be just one memcpy.